File: app/src/main/java/com/example/all_sensors/presentation/PpgListener.java
package com.example.all_sensors.presentation;

import android.util.Log;

import com.example.all_sensors.presentation.data.DataBuffer;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.data.DataPoint;
import com.samsung.android.service.health.tracking.data.ValueKey;

import java.util.List;
import java.util.Locale;
public class PpgListener implements HealthTracker.TrackerEventListener{
    private final DataBuffer buffer;
    public PpgListener(DataBuffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public void onDataReceived(List<DataPoint> dataPoints) {
        long now = System.currentTimeMillis();
        for (com.samsung.android.service.health.tracking.data.DataPoint dp : dataPoints) {
            float green = dp.getValue(ValueKey.PpgSet.PPG_RED);
            float ir    = dp.getValue(ValueKey.PpgSet.PPG_IR);
            float red   = dp.getValue(ValueKey.PpgSet.PPG_RED);

            Log.d("PPG_DATA", String.format(Locale.US, "PPG @%d → G:%.1f IR:%.1f R:%.1f",
                    now, green, ir, red));

            String vals = String.format(Locale.US, "%.2f,%.2f,%.2f", green, ir, red);
            buffer.add(new com.example.all_sensors.presentation.data.DataPoint(now, "PPG", vals));

            ObserverUpdater.getObserverUpdater().notifyAccObservers(green,ir,red);
        }
    }

    @Override public void onFlushCompleted() { }
    @Override public void onError(com.samsung.android.service.health.tracking.HealthTracker.TrackerError error) {
        Log.e("PPG_DATA", "PPG error");
    }
}

File: app/src/main/java/com/signals/smartwatch/sensors/samsung/PpgListener.kt
package com.signals.smartwatch.sensors.samsung

import android.util.Log
import com.signals.smartwatch.data.SensorDao
import com.signals.smartwatch.data.SensorEntity
import com.samsung.android.service.health.tracking.HealthTracker
import com.samsung.android.service.health.tracking.data.DataPoint
import com.samsung.android.service.health.tracking.data.ValueKey

class PpgListener(
    private val dao: SensorDao,
    private val events: com.signals.smartwatch.SensorEvents
) : HealthTracker.TrackerEventListener {

    override fun onDataReceived(dataPoints: List<DataPoint>) {
        val now = System.currentTimeMillis()
        dataPoints.forEach { dp ->
            val green = dp.getValue(ValueKey.PpgSet.PPG_GREEN)
            val ir    = dp.getValue(ValueKey.PpgSet.PPG_IR)
            val red   = dp.getValue(ValueKey.PpgSet.PPG_RED)
            Log.d(TAG, "PPG @$now → G:${"%.1f".format(green)} IR:${"%.1f".format(ir)} R:${"%.1f".format(red)}")
            dao.insertBlocking(SensorEntity(
                timestamp = now,
                sensor    = "PPG",
                values    = "%.2f,%.2f,%.2f".format(green, ir, red)
            ))
            events.onPpg(green.toFloat(), ir.toFloat(), red.toFloat())
        }
    }

    override fun onFlushCompleted() = Unit

    override fun onError(error: HealthTracker.TrackerError) {
        Log.e(TAG, "Error: $error")
    }

    companion object { private const val TAG = "PPG_DATA" }
}
